SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp

```diff
@@ -70,3 +74,80 @@ int32 USpatialNetConnection::IsNetReady(bool Saturate)
 	// This will always return true until we solve this.
 	return true;
 }
+
+void USpatialNetConnection::InitHeartbeat(FTimerManager* InTimerManager, Worker_EntityId InPlayerControllerEntity)
+{
+	checkf(PlayerControllerEntity == SpatialConstants::INVALID_ENTITY_ID, TEXT("InitHeartbeat: PlayerControllerEntity already set: %lld. New entity: %lld"), PlayerControllerEntity, InPlayerControllerEntity);
+	PlayerControllerEntity = InPlayerControllerEntity;
+	TimerManager = InTimerManager;
+
+	bIsServer = Driver->IsServer();
+
+	if (Driver->IsServer())
+	{
+		SetHeartbeatTimeoutTimer();
+
+		// Set up heartbeat event callback
+		TWeakObjectPtr<USpatialNetConnection> ConnectionPtr = this;
+		Cast<USpatialNetDriver>(Driver)->Receiver->AddHeartbeatDelegate(PlayerControllerEntity, HeartbeatDelegate::CreateLambda([ConnectionPtr](Worker_ComponentUpdateOp& Op)
+		{
+			if (ConnectionPtr.IsValid())
+			{
+				Schema_Object* EventsObject = Schema_GetComponentUpdateEvents(Op.update.schema_type);
+				uint32 EventCount = Schema_GetObjectCount(EventsObject, SpatialConstants::HEARTBEAT_EVENT_ID);
+				if (EventCount > 0)
+				{
+					if (EventCount > 1)
+					{
+						UE_LOG(LogTemp, Warning, TEXT("Received multiple heartbeat events in a single component update, entity %lld."), ConnectionPtr->PlayerControllerEntity);
```

Comment:
Is this bad? I feel like this is very likely to happen due to the runtime automatically squishing events together.